(DOCSP-24576): Flutter open realm async (#2236)
mongodb/docs-realm: source/sdk/flutter/sync/open-synced-realm.txt

<flutter-manage-sync-session>`.

.. _flutter-open-realm-async:

Sync Data on Open
~~~~~~~~~~~~~~~~~

To sync all data with App Services when you open a realm, use the asynchronous method
:flutter-sdk:`Realm.open() <realm/Realm/open.html>`. The operation syncs all available
data before returning the ``Realm``.

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.async-open.dart
   :language: dart

To track the state of the synchronization, pass a :flutter-sdk:`ProgressCallback <realm/ProgressCallback.html>`
to the optional named argument ``onProgressCallback``.

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.async-open-track-progress.dart
   :language: dart

To be able to cancel a synchronization in progress, pass a `CancellationToken
<https://pub.dev/documentation/cancellation_token/latest/cancellation_token/CancellationToken-class.html>`__
instance to the optional named argument ``cancellationToken``.
Call `CancellationToken.cancel() <https://pub.dev/documentation/cancellation_token/latest/cancellation_token/CancellationToken/cancel.html>`__

> to cancel the synchronization.

Reviewer: In addition, it is also possible to cancel more than one opening realm async operations with the came cancellation token.

Author: i don't think that needs inclusion in the docs. this wouldn't be a very common use case, would it? 

as long as we're not saying this is forbidden, i think it's ok as is. thoughts?

Reviewer 2: Agree - I think docs should avoid getting too much into technicalities and cover the most common use cases.

Reviewer: Ok. I've just mentioned it.

:flutter-sdk:`Configuration.flexibleSync() <realm/Configuration/flexibleSync.html>` constructor.
This constructor returns a :flutter-sdk:`FlexibleSyncConfiguration <realm/FlexibleSyncConfiguration-class.html>`.
Then, pass the ``FlexibleSyncConfiguration`` to :flutter-sdk:`Realm() <realm/Realm-class.html>`
to open an instance of the realm. Data syncs with App Services in the background
after you open the realm.

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.open-flexible-sync-realm.dart
   :language: dart

Once you open a synced realm, :ref:`configure and manage the sync session
<flutter-manage-sync-session>`.

.. _flutter-open-realm-async:

> Sync Data on Open

Reviewer 3: I wonder if we might want to rename this section to something that makes it clearer that you're opening the realm _after_ downloading changes. I wouldn't necessarily know from this heading that this will block opening the realm. Maybe something like "Open a Realm After Downloading Changes" or similar?

~~~~~~~~~~~~~~~~~

To sync all data with App Services when you open a realm, use the asynchronous method
:flutter-sdk:`Realm.open() <realm/Realm/open.html>`. The operation syncs all available

> data before returning the ``Realm``. ``Realm.open()`` only resolves if the app

Reviewer 3: It's very similar to what you already have here, but when the Swift SDK had a similar method, some of our docs readers found it clearer to say something like "If the user is offline, Realm.open() does not resolve" vs. something like "Realm.open() only resolves if the app is online." Make it explicit that this method does not work when offline. This also speaks more directly to our positioning Realm w/Sync as an "offline-first DB."

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.async-open.dart
   :language: dart

To track the state of the synchronization, pass a :flutter-sdk:`ProgressCallback <realm/ProgressCallback.html>`
to the optional named argument ``onProgressCallback``.

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.async-open-track-progress.dart
   :language: dart

To be able to cancel a synchronization in progress, pass a `CancellationToken
<https://pub.dev/documentation/cancellation_token/latest/cancellation_token/CancellationToken-class.html>`__
instance to the optional named argument ``cancellationToken``.
Call `CancellationToken.cancel() <https://pub.dev/documentation/cancellation_token/latest/cancellation_token/CancellationToken/cancel.html>`__
to cancel the synchronization.

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.async-open-cancel.dart
   :language: dart

.. example:: Realm() vs. Realm.open()

   This section compares examples of scenarios when you might want to use ``Realm()``
   versus ``Realm.open()`` to open a realm in an application.

>    Say you have an app that allowed the user to record and save their

Reviewer 3: Nit: "Say" is idiomatic here, and may not be the best phrasing for people outside the US or people for whom English is a second language. We might instead want to say something like "Consider an app that allows" or "For example, your app might allow the user..."

Same for the paragraph down below.

Author: absolutely.

btw this phrasing was taken from the legacy swift page https://www.mongodb.com/docs/realm/sdk/swift/sync/sync-realm-open-legacy/#open-a-synced-realm-after-downloading-changes

maybe not worth making changes to legacy page, but wanted to put on your radar.

Reviewer 3: Hah, what was I thinking? 😉 

I am trying very hard not to touch that page ever again, so I'll leave it for now, but thanks for the heads-up!

Then, pass the ``FlexibleSyncConfiguration`` to :flutter-sdk:`Realm() <realm/Realm-class.html>`
to open an instance of the realm. Data syncs with App Services in the background
after you open the realm.

.. literalinclude:: /examples/generated/flutter/open_flexible_sync_realm_test.snippet.open-flexible-sync-realm.dart
   :language: dart

Once you open a synced realm, :ref:`configure and manage the sync session
<flutter-manage-sync-session>`.

.. _flutter-open-realm-async:

Open a Realm After Downloading Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To sync all data with App Services when you open a realm, use the asynchronous method
:flutter-sdk:`Realm.open() <realm/Realm/open.html>`. The operation syncs all available
data before returning the realm.

On first open, ``Realm.open()`` downloads all data that matches your sync subscriptions.
Subsequent opens only download the latest changes.
Depending on initial realm size and updates to the data set while the device is not syncing,
performance may be slower on first open and faster on subsequent opens.

If the client application is offline, ``Realm.open()`` does not resolve.

> You should check if the device is connected to the internet before using ``Realm.open()``.

Reviewer: ```suggestion
You could check if the device is connected to the internet before using ``Realm.open()`` or just to cancel the operation ``Realm.open()`` after some timeout.
```

> If the device is not connected to the internet, you can still use ``Realm()``

Reviewer: ```suggestion
If the device is not connected to the internet, you can always open your local realm using ``Realm()``
```

> to open a realm and sync data in the background when an internet connection is available.

Reviewer: ```suggestion
The data will be automatically synced in background once the  internet connection is available.
```
